### PR TITLE
✨(backend) make LiveKit Egress recording encoding configurable

### PR DESCRIPTION
**Pull request by damienlaine**
## Purpose

Recordings produced by LiveKit Egress currently use the hard-coded `H264_720P_30` preset (1280×720 / 30 fps / 3000 kbps H.264 MAIN + 128 kbps AAC). That's ~690 MB per 30-minute meeting. For typical La Suite Meet workloads (talking heads + screen share, low motion) this is significantly heavier than necessary, and the fixed 30 fps / veryfast real-time encoding also keeps egress worker CPU higher than it needs to be.

Operators have no way to tune this today without patching the backend.

## Proposal

Expose seven `RECORDING_ENCODING_*` settings that, when `RECORDING_ENCODING_ENABLED=True`, are sent to LiveKit as `EncodingOptions.advanced` in `RoomCompositeEgressRequest`. Default: disabled → current behaviour unchanged.

Values pass straight through to the GStreamer pipeline (`x264enc` / `faac`), no hidden conversions. Lowering `FRAMERATE` is the main CPU lever (x264 `speed-preset` stays `veryfast` on the LiveKit side); lowering `VIDEO_BITRATE_KBPS` is the main file-size lever.

Documentation includes a GStreamer mapping table and a set of reference profiles (Default / Balanced / Low CPU / Minimum CPU / Audio-heavy) with estimated sizes and CPU cost. The recommended "Low CPU / small file" profile lands at ~150 MB for 30 min with ~50 % of the default CPU load.

- [x] Add `RECORDING_ENCODING_*` settings (opt-in, defaults match the current `H264_720P_30` preset)
- [x] Pass `advanced` `EncodingOptions` only when enabled, so disabled-path behaviour is byte-identical
- [x] Wire the new values through `WorkerServiceConfig.from_settings()`
- [x] Unit tests for both opt-in and opt-out paths (services + factories)
- [x] `env.d/development/common.dist` documented examples
- [x] `docs/features/recording.md` — configuration table + "Tuning recording encoding" section with GStreamer mapping and reference-profile table
- [x] CHANGELOG entry under `[Unreleased]`

### CODE REVIEW

File: docs/features/recording.md [hunk 1]
```diff
@@ -141,3 +148,53 @@ Using default project meet
...
+| ---------------------- | ---------- | --- | ------------ | ------------ | ------------ | --------------- | ------------------------ | --------------------------------------------------- |
+| Default (preset)       | 1280×720   | 30  | 3000         | 128          | 4            | **~690 MB**     | 100 %                    | Unchanged LiveKit behaviour                         |
+| Balanced               | 1280×720   | 20  | 1000         | 96           | 6            | ~240 MB         | ~67 %                    | Mixed content, moderate motion                      |
+| **Low CPU / small file** | 1280×720 | 15  | 600          | 64           | 10           | **~150 MB**     | ~50 %                    | Videoconference (talking heads + screen share) ★    |
+| Minimum CPU            | 960×540    | 15  | 500          | 64           | 10           | ~125 MB         | ~30 %                    | Voice-first meetings, readable text not required    |
+| Audio-heavy fallback   | 1280×720   | 10  | 400          | 96           | 15           | ~110 MB         | ~35 %                    | Long webinars, low motion                           |
```
Comment (lebaudantoine) [inline docs/features/recording.md]: how were computed the Egress CPU?

Comment (lebaudantoine) [inline docs/features/recording.md]: Are 6+ (s) keyframes a bit aggressive for a meeting recording that's usually replayed?

File: docs/features/recording.md [hunk 2]
```diff
@@ -141,3 +148,53 @@ Using default project meet
...
+
+★ Recommended starting point for typical La Suite Meet usage.
```
Comment (lebaudantoine) [inline docs/features/recording.md]:
```suggestion
★ Recommended starting point for typical LaSuite Meet usage.
```

File: docs/features/recording.md [hunk 3]
```diff
@@ -141,3 +148,53 @@ Using default project meet
...
+
+### Reference profiles
+
+Rough 30-minute file-size estimates assume video + audio bitrate multiplied by duration. Actual sizes vary with content (static talking heads compress better than heavy screen motion).
+
+| Profile                | Resolution | FPS | Video (kbps) | Audio (kbps) | Keyframe (s) | ~ size / 30 min | Egress CPU (vs. default) | Suitable for                                        |
+| ---------------------- | ---------- | --- | ------------ | ------------ | ------------ | --------------- | ------------------------ | --------------------------------------------------- |
+| Default (preset)       | 1280×720   | 30  | 3000         | 128          | 4            | **~690 MB**     | 100 %                    | Unchanged LiveKit behaviour                         |
+| Balanced               | 1280×720   | 20  | 1000         | 96           | 6            | ~240 MB         | ~67 %                    | Mixed content, moderate motion                      |
+| **Low CPU / small file** | 1280×720 | 15  | 600          | 64           | 10           | **~150 MB**     | ~50 %                    | Videoconference (talking heads + screen share) ★    |
```
Comment (lebaudantoine) [inline docs/features/recording.md]: Your recommendation has a contradiction; this preset would result in blurry screen-share, impacting the readability of slides shared. Is it really a good starting point?

File: src/backend/core/recording/worker/services.py [hunk 4]
```diff
@@ -83,6 +83,28 @@ def start(self, room_name, recording_id):
...
+        return livekit_api.EncodingOptions(
+            width=opts["width"],
+            height=opts["height"],
+            framerate=opts["framerate"],
+            video_codec=livekit_api.VideoCodec.H264_MAIN,
+            video_bitrate=opts["video_bitrate"],
+            audio_codec=livekit_api.AudioCodec.AAC,
+            audio_bitrate=opts["audio_bitrate"],
+            audio_frequency=48000,
+            key_frame_interval=opts["key_frame_interval"],
```
Comment (lebaudantoine) [inline src/backend/core/recording/worker/services.py]: This mixes hardcoded values, LiveKit defaults explicitly passed, and values that can be overridden by Django settings.

File: src/backend/meet/settings.py [hunk 5]
```diff
@@ -700,6 +700,41 @@ class Base(Configuration):
...
+    RECORDING_ENCODING_AUDIO_BITRATE_KBPS = values.PositiveIntegerValue(
+        128,
+        environ_name="RECORDING_ENCODING_AUDIO_BITRATE_KBPS",
+        environ_prefix=None,
+    )
+    RECORDING_ENCODING_KEY_FRAME_INTERVAL_S = values.FloatValue(
+        4.0,
+        environ_name="RECORDING_ENCODING_KEY_FRAME_INTERVAL_S",
+        environ_prefix=None,
+    )
```
Comment (lebaudantoine) [inline src/backend/meet/settings.py]: Reading the code, it looks like we should update the inline comment to explicitly state that only composite egress is affected by these settings, and that audio-only recordings remain unchanged.

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,6 +11,10 @@ and this project adheres to
 ### Added
 
 - ✨(backend) add metadata collection of VAD, connection and chat events
+- ✨(backend) make LiveKit Egress recording encoding configurable via
+  `RECORDING_ENCODING_*` settings (disabled by default, falls back to the
+  H264_720P_30 preset). Lowering framerate and bitrate reduces recording
+  file size and egress worker CPU load.
```
Comment (lebaudantoine) [inline CHANGELOG.md]: doesn't align with other changelog entries

**Review by lebaudantoine — COMMENTED** [review]
Thank you for your contribution, I haven’t tested it locally yet, but it looks promising.

Feel free to add more tests if any additional edge cases come to mind. We’re getting very close to something that’s ready to be merged.

On a related note, I’d like to allow users to pass recording encoding configuration from the frontend so they can customize their recording preferences. I’ll start working on that once your PR is merged.

**Comment by lebaudantoine** [comment]
@coderabbitai full review

File: docs/features/recording.md (hunk 1 above)
Comment (damienlaine) [reply docs/features/recording.md]: Indicative measurements on one core of a Ryzen laptop saturated by the default preset (= 100 %). Added a note to the table in 87b04ed: scaling is roughly linear in `framerate × bitrate` but absolute numbers depend on the host hardware — we'd need a real benchmark on the actual egress host for production figures.

Comment (damienlaine) [reply docs/features/recording.md]: Good catch — aligned all profiles to 4s (standard VOD value). Side note: I also fixed an inaccurate claim in the docs that said larger keyframe intervals reduce file size by 10–15 %; in ABR mode (which LiveKit Egress uses) the bitrate budget is fixed, so keyframe interval only trades seek granularity for slightly higher quality at the same bitrate.

File: docs/features/recording.md (hunk 2 above)
Comment (damienlaine) [reply docs/features/recording.md]: Applied.

File: docs/features/recording.md (hunk 3 above)
Comment (damienlaine) [reply docs/features/recording.md]: Looked into this — the analysis is right but the conclusion was framerate-dependent. The relevant metric for text legibility on screen content is bits/frame, not absolute bitrate. At 720p the threshold where dense text starts to break down is around ~40 kbits/frame:

- 600 kbps × 30 fps = 20 kbits/frame → blurry slides (the case the caveat was warning about)
- 600 kbps × 15 fps = 40 kbits/frame → borderline OK (our preset)
- 900 kbps × 15 fps = 60 kbits/frame → comfortable for dense slides

So the ★ Low CPU preset (600 kbps × 15 fps) sits exactly at the threshold — fine for talking-head-dominant meetings with occasional slides, but tight for deck-heavy sessions. I've:

- Reframed the caveat around bits/frame so the threshold reasoning is explicit, and noted that **lowering framerate is a more screen-share-friendly lever than lowering bitrate** (since dropping fps multiplies bits/frame proportionally and slide content has very low motion temporally anyway).
- Added a 'Slide-heavy' reference profile (900 kbps × 15 fps) for IDE/deck-share meetings, and tightened the ★ Low CPU 'Suitable for' to 'talking-head dominant + occasional slides'.

File: src/backend/core/recording/worker/services.py (hunk 4 above)
Comment (damienlaine) [reply src/backend/core/recording/worker/services.py]: Refactored — all the EncodingOptions kwargs (operator-tunable values + pinned codec/frequency constants) are now assembled in `WorkerServiceConfig.from_settings()`, and the services layer is a thin protobuf adapter (`EncodingOptions(**opts)`). Single source of truth, no more mix at the call site.

File: src/backend/meet/settings.py (hunk 5 above)
Comment (damienlaine) [reply src/backend/meet/settings.py]: Done — the inline comment now explicitly states that these settings affect `VideoCompositeEgressService` only and are silently ignored by `AudioCompositeEgressService` (audio-only transcript recordings, whose request never carries advanced EncodingOptions).

File: CHANGELOG.md (hunk 6 above)
Comment (damienlaine) [reply CHANGELOG.md]: Aligned — collapsed to a single-line entry matching the surrounding style. (The entry also moved out of 1.15.0 into [Unreleased] when I rebased on latest main.)

**Comment by damienlaine** [comment]
Thanks for the review @lebaudantoine! Pushed [87b04ed](https://github.com/suitenumerique/meet/pull/1288/commits/87b04ed609264e76ad6311e94d4263896ce2dcbc) (force-push, rebased onto latest main) addressing every thread:

- **Screen-share contradiction**: reframed the caveat around `bits/frame` (the actual driver of text legibility on screen content), kept the ★ Low CPU preset for talking-head dominant meetings, added a new **Slide-heavy** reference profile (900 kbps × 15 fps) for deck/IDE-share heavy use cases. Lowering framerate is the safer screen-share lever than lowering bitrate — explained inline.
- **Keyframe interval**: aligned all profiles to 4s (standard VOD value). Also fixed an inaccurate doc claim about file-size reduction (LiveKit Egress runs x264 in ABR, so keyframe interval doesn't shift the bitrate budget).
- **CPU figures**: noted the methodology (single Ryzen laptop core saturated by default preset, ~linear scaling) — these are indicative, not benchmarked on egress hosts.
- **services.py mix**: refactored — `WorkerServiceConfig.from_settings()` now assembles the full EncodingOptions kwargs dict (operator-tunable values + pinned codec/freq constants); the services layer is just `EncodingOptions(**opts)`.
- **Settings audio-only scope**: comment updated to explicitly state these settings affect `VideoCompositeEgressService` only and are silently ignored by `AudioCompositeEgressService`.
- **CHANGELOG**: collapsed to single-line entry, moved to `[Unreleased]` after rebase.
- Trivial suggestion + previous force-push merge conflict also resolved.

Tests `core/tests/recording/worker/{test_factories,test_services}.py` pass locally (26/26).

File: src/backend/core/recording/worker/factories.py [hunk 7]
```diff
@@ -8,8 +8,17 @@
...
+
 logger = logging.getLogger(__name__)
 
+# Codec / frequency constants matching LiveKit's H264_720P_30 preset.
+# Kept fixed because changing them would shift the goal-post away from the
+# "safe drop-in replacement for the default preset" contract of this feature.
+_RECORDING_VIDEO_CODEC = livekit_api.VideoCodec.H264_MAIN
+_RECORDING_AUDIO_CODEC = livekit_api.AudioCodec.AAC
+_RECORDING_AUDIO_FREQUENCY_HZ = 48000
+
```
Comment (lebaudantoine) [inline src/backend/core/recording/worker/factories.py:23]: +1

**Review by lebaudantoine — APPROVED** [review]
LGTM thank @damienlaine

**Comment by lebaudantoine** [comment]
@coderabbitai full review

File: src/backend/core/tests/recording/worker/test_services.py [hunk 8]
```diff
@@ -224,6 +249,40 @@ def test_video_composite_egress_start_missing_egress_id(video_service):
...
+    advanced = request.advanced
+    assert advanced.width == 1280
+    assert advanced.height == 720
+    assert advanced.framerate == 15
+    assert advanced.video_bitrate == 600
+    assert advanced.audio_bitrate == 64
+    assert advanced.key_frame_interval == pytest.approx(10.0)
+    assert advanced.video_codec == livekit_api.VideoCodec.H264_MAIN
+    assert advanced.audio_codec == livekit_api.AudioCodec.AAC
+
```
[the remark this answers is not in the document]
Comment (damienlaine) [reply src/backend/core/tests/recording/worker/test_services.py:285]: Good catch — added `assert advanced.audio_frequency == 48000` in 61d68b6. Tests pass locally (26/26).
